### PR TITLE
Make immutable downloads less chatty

### PR DESCRIPTION
**Pull request by itamarst**
When latency is non-zero (i.e. anything not a LAN) this has huge improvement to download times, with both HTTP and Foolscap.

Downside: on Foolscap this increases CPU usage somewhat. Still seems worth it, presumably most Tahoe-LAFS interactions are not over LAN.

Fixes https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3946

### CODE REVIEW

File: src/allmydata/immutable/downloader/node.py [hunk 1]
```diff
@@ -76,7 +78,7 @@ def __init__(self, verifycap, storage_broker, secret_holder,
         # .guessed_segment_size, .guessed_num_segments, and
         # .ciphertext_hash_tree (with a dummy, to let us guess which hashes
         # we'll need)
-        self._build_guessed_tables(DEFAULT_MAX_SEGMENT_SIZE)
+        self._build_guessed_tables(self.default_max_segment_size)
```
Comment (exarkun) [inline src/allmydata/immutable/downloader/node.py:81]: Uhh this is so weird.  Guess a value that is trivially readable from the front of any share and derive a bunch of real values from it -- then replace all those values with the real values after doing the trivial read -- supposedly as an optimization for `Segmentation` -- which cannot possibly work correctly if you guessed wrong? :/ :/ :/ :/

I guess I get to read some more of `allmydata/immutable/downloader/` now...

File: src/allmydata/test/test_system.py [hunk 2]
```diff
@@ -1811,6 +1814,45 @@ def _check_lit_filenode_results(r):
         d.addCallback(_got_lit_filenode)
         return d
 
+    @async_to_deferred
+    async def test_upload_download_immutable_different_default_max_segment_size(self):
```
Comment (exarkun) [inline src/allmydata/test/test_system.py:1818]: Perhaps the integration test suite is better suited to this kind of thing?

**Review by exarkun — COMMENTED** [review]
Thanks.  I think I have to learn some more stuff before I can give an approving review.

File: integration/util.py [hunk 3]
```diff
@@ -46,6 +46,7 @@
     write_config,
 )
 from allmydata import client
+from allmydata.interfaces import DEFAULT_MAX_SEGMENT_SIZE
```
[the remark this answers is not in the document]
Comment (itamarst) [reply integration/util.py:49]: That matches my understanding, yes.

File: src/allmydata/immutable/downloader/node.py (hunk 1 above)
Comment (itamarst) [reply src/allmydata/immutable/downloader/node.py:81]: If it can't work if you guess wrong then this PR is no good :(

Comment (itamarst) [reply src/allmydata/immutable/downloader/node.py:81]: But if you can get the segment value from trivial read... I should do that instead?

File: src/allmydata/test/test_system.py (hunk 2 above)
Comment (itamarst) [reply src/allmydata/test/test_system.py:1818]: Hmhm. I was thinking I'd need to push the segment size through another N layers of abstraction but I guess with a config option then maybe not?

**Comment by itamarst** [comment]
Note that the integration test failure with i2pd will be fixed when #1250 is merged.

File: src/allmydata/immutable/downloader/node.py (hunk 1 above)
Comment (exarkun) [reply src/allmydata/immutable/downloader/node.py:81]: It seems like incorrect guesses don't impair functionality.  Maybe they make performance slightly worse or something (we could try to measure but I don't particularly want to at the moment).

So I think this part is fine as-is.

**Review by exarkun — APPROVED** [review]
This looks good to me (in particular, I can't see how it would break compatibility with existing clients or stored data - and the new integration test seems to support this conclusion pretty well).

**Pull request merged**